test_progs/pcq.c

```diff
@@ -0,0 +1,1066 @@
+	       "\n", progname, progname, progname, progname, progname, progname);
+}
+
+int
+main(int argc, char **argv)
+{
+	pthread_t producer_thread, consumer_thread;
+	struct pcq_thread_arg prod = { 0 };
+	struct pcq_thread_arg cons = { 0 };
+	char *statusfname = NULL;
+	char *filename = NULL;
+	bool producer = false;
+	bool consumer = false;
+	bool create = false;
+	u64 bucket_size = 0;
+	bool drain = false;
+	u64 nmessages = 0;
+	bool info = false;
+	FILE *statusfile;
+	u64 nbuckets = 0;
+	int wait = true;
+	int runtime = 0;
+	int verbose = 0;
+	s64 seed = 0;
+	int arg_ct;
+	int c, rc;
+	s64 mult;
+
+	struct option pcq_options[] = {
+		/* These options set a flag. */
+		{"bsize",       required_argument,        0,  'b'},
+		{"nbuckets",    required_argument,        0,  'n'},
+		{"seed",        required_argument,        0,  'S'},
+		{"nmessages",   required_argument,        0,  'N'},
+		{"statusfile",  required_argument,        0,  'f'},
+		{"time",        required_argument,        0,  't'},
+
+		{"create",      no_argument,              0,  'C'},
+		{"producer",    no_argument,              0,  'p'},
+		{"consumer",    no_argument,              0,  'c'},
+		{"info",        no_argument,              0,  'i'},
+		{"drain",       no_argument,              0,  'd'},
+		{"dontflush",   no_argument,              0,  'D'},
+		/* These options don't set a flag.
+		 * We distinguish them by their indices.
+		 */
+		/*{"dryrun",       no_argument,       0, 'n'}, */
+		{0, 0, 0, 0}
+	};
+
+	/* Note: the "+" at the beginning of the arg string tells getopt_long
+	 * to return -1 when it sees something that is not recognized option
+	 * (e.g. the command that will mux us off to the command handlers
+	 */
+	while ((c = getopt_long(argc, argv, "+b:s:S:n:N:f:t:CdpcwDih?v",
+				pcq_options, &optind)) != EOF) {
+		char *endptr;
+
+		arg_ct++;
+		switch (c) {
+
+		case 'i':
+			info = true;
+			break;
+
+			/* Create operations */
+		case 'C':
+			create = true;
+			break;
+
+		case 'b':
+			bucket_size = strtoull(optarg, &endptr, 0);
+			mult = get_multiplier(endptr);
+			if (mult > 0)
+				bucket_size *= mult;
+			printf("bucket_size=%lld\n", bucket_size);
+			break;
+
+		case 'n':
+			nbuckets = strtoull(optarg, &endptr, 0);
+			mult = get_multiplier(endptr);
+			if (mult > 0)
+				nbuckets *= mult;
+			printf("nbuckets=%lld\n", nbuckets);
+			break;
+
+		case 'S':
+			seed = strtoull(optarg, 0, 0);
+			break;
+
+		case 't':
+			runtime = strtoull(optarg, 0, 0);
+			break;
+
+		case 'p':
+			producer = true;
+			break;
+
+		case 'c':
+			consumer = true;
+			break;
+
+		case 'd':
+			drain = true;
+			wait = false;
+			break;
+
+		case 'v':
+			verbose++;
+			break;
+
+		case 'N':
+			nmessages = strtoull(optarg, &endptr, 0);
+			mult = get_multiplier(endptr);
+			if (mult > 0)
+				nmessages *= mult;
+			break;
+			break;
+
+		case 'f':
+			/* Write execution status to this file (for testing) */
+			statusfname= optarg;
```

Comment:
In pcq CLI, if --statusfile option is not given, it hits a segmentaiton fault at line 1062. 
========================
$ sudo ./pcq --producer --consumer -t 5 /mnt/famfs/queue13

Queue:    /mnt/famfs/queue13
Producer: nsent=701286 nerrors=0
Consumer: nreceived=701286 nerrors=0
Segmentation fault

=================================
It seems like when it does a fopen it is not identifying that file is not provided..
We need to make it a mandatory field or handle the case where it is not provided by user.

Reply:
Actually, the problem was that "FILE *statusfile = NULL" was not initialized to null (but now it is). Smoke uses the --statusfile option to get detailed counts from pcq that don't fit in a single byte exit code.

Valgrind found that this morning...